examples/misc/navigate_search_result_pages_caching_tokens.rb

```diff
@@ -0,0 +1,186 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# GoogleAdsService.Search results are paginated but they can only be retrieved
+# in sequence starting by the first page. More details at
+# https://developers.google.com/google-ads/api/docs/reporting/paging.
+#
+# This example searches campaigns illustrating how GoogleAdsService.Search
+# result page tokens can be cached and reused to retrieve previous pages. This
+# is useful when you need to request pages that were already requested in the
+# past without starting over from the first page. For example, it can be used
+# to implement an interactive application that displays a page of results at a
+# time without caching all the results first.
+#
+# To add campaigns, run add_campaigns.rb.
+
+require 'optparse'
+require 'google/ads/google_ads'
+
+def navigate_search_result_pages_caching_tokens(customer_id)
+  # GoogleAdsClient will read a config file from
+  # ENV['HOME']/google_ads_config.rb when called without parameters
+  client = Google::Ads::GoogleAds::GoogleAdsClient.new
+
+  # The cache of page tokens. The key is the 0-indexed page number.
+  # The first page's token is always the empty string.
+  page_tokens = {
+    0 => ''
+  }
+
+  puts "Fetching page 1 to get metadata"
+  query = "SELECT campaign.id, campaign.name FROM campaign ORDER BY campaign.name LIMIT #{RESULTS_LIMIT}"
+  response = client.service.google_ads.search(
+    customer_id: customer_id,
+    query: query,
+    page_size: PAGE_SIZE,
+    return_total_results_count: true,
+  )
+
+  cache_next_page_token(page_tokens, response.page, 1)
+
+  # The total results count does not take into consideration the LIMIT clause
+  # of the query so we need to find the minimal value between the limit and the
+  # total results count.
+  total_results = [
+    RESULTS_LIMIT,
+    response.page.response.total_results_count,
+  ].min
+  puts "Found ${total_results} campaigns."
+
+  total_pages = (total_results.to_f / PAGE_SIZE).ceil
+  raise "Could not find any campaigns." if total_pages == 0
+
+  middle_page = (total_pages / 2.0).ceil
+  puts "Printing results from page #{middle_page}."
+  fetch_and_print_page_results(
+    client,
+    customer_id,
+    query,
+    PAGE_SIZE,
+    true,
+    middle_page,
+    page_tokens,
+  )
+
+  puts 'Print all pages, starting at the last page and moving to the first.'
+  (1..total_pages).reverse_each do |page_number|
+    fetch_and_print_page_results(
+      client,
+      customer_id,
+      query,
+      PAGE_SIZE,
+      true,
+      page_number,
+      page_tokens,
+    )
+  end
+end
+
+# [START navigate_search_result_pages_caching_tokens]
+def fetch_and_print_page_results(client, customer_id, query, page_size,
+                                 return_total_results_count, page_number, page_tokens)
+  if page_tokens.has_key?(page_number)
```

Comment:
Should not `page_number` be `page_number  - 1`?

Reply:
Good catch, thanks! It *looked* like it was working with this mistake but there was a subtle error.